tests/integration/conftest.py

```diff
@@ -69,3 +69,23 @@ def local_pypiserver() -> Iterator[None]:
             raise RuntimeError("pypiserver didn't start fast enough")
 
         yield
+
+
+def does_not_test_any_affected_pms(test: Any, affected_pms_to_test: set[str]) -> bool:
+    pm_under_test = Path(test.fspath).stem.lstrip("test_")
+    return pm_under_test not in affected_pms_to_test
+
+
+def pytest_collection_modifyitems(session: Any, config: Any, items: Any) -> None:
+    skip_mark = pytest.mark.skip(reason="No changes to tested code")
+    to_test = utils.determine_integration_tests_to_run_for("HEAD")
+    if to_test is utils.TEST_EVERYTHING:
+        return
+    elif to_test is utils.TEST_NOTHING:
+        for item in items:
+            item.add_marker(skip_mark)
+    else:
+        for item in items:
+            if does_not_test_any_affected_pms(item, to_test):
+                item.add_marker(skip_mark)
+    return
```

Comment:
Apparently, you can use marks (as you clearly demonstrate), but that means that after this function is run, pytest still evaluates **every single test** and skips it on demand based on the mark. I'm not saying that's wrong by any means, I'll just share what I envisioned and let others to comment on that.

```
...
if to_test is utils.TEST_NOTHING:
  items[:] = []  # items are passed by reference, so need to be modified in-place
else:
  deselected = {i for i in items if does_not_test_any_affected_pms(i, to_test)}
  items[:] = set(items) - deselected
  
 # inform pytest of the deselected tests
 config.hook.pytest_deselected(items=deselected)


def pytest_deselected(items):
# override the built-in method to print deselected tests (mostly for verification purposes that this machinery works)
print("Deselected tests:")
for i in items:
  print(f"  - {i.nodeid}")
```

Here's the output comparison:

**Yours**:
```
collected 51 items    <<< NOTICE THIS LINE                                                             

tests/integration/test_gomod.py ssssssssssssssss
tests/integration/test_npm.py ssssssss
...
```

**Mine**:
```
collecting ... 
Deselected tests:
  - tests/integration/test_gomod.py::test_e2e_gomod[gomod_e2e_multiple_modules]
  - tests/integration/test_yarn.py::test_yarn_packages[yarn_incorrect_checksum]
  - tests/integration/test_rpm.py::test_repo_files[rpm_test_repo_file]
  - tests/integration/test_gomod.py::test_gomod_packages[gomod_without_deps]
  - tests/integration/test_yarn.py::test_yarn_packages[yarn_disallowed_protocols]
  - tests/integration/test_gomod.py::test_e2e_gomod[gomod_1.21_e2e_test]
  - tests/integration/test_gomod.py::test_gomod_packages[gomod_vendored_with_flag]
  - tests/integration/test_gomod.py::test_gomod_packages[gomod_vendor_check_wrong_vendor]
  - tests/integration/test_rpm.py::test_rpm_packages[rpm_multiple_archs]
  - tests/integration/test_yarn.py::test_yarn_packages[yarn_immutable_installs]
  - tests/integration/test_npm.py::test_npm_smoketest[npm_lockfile3_yarn_registry]
  - tests/integration/test_yarn.py::test_yarn_packages[yarn_no_lockfile]
  - tests/integration/test_gomod.py::test_gomod_packages[gomod_vendor_check_empty_vendor]
  - tests/integration/test_npm.py::test_e2e_npm[npm_smoketest_lockfile3]
  - tests/integration/test_gomod.py::test_gomod_packages[gomod_workspaces]
  - tests/integration/test_gomod.py::test_gomod_packages[gomod_go_generate_imported]
  - tests/integration/test_gomod.py::test_gomod_packages[gomod_vendored_without_flag]
  - tests/integration/test_gomod.py::test_gomod_packages[gomod_vendor_check_no_vendor]
  - tests/integration/test_npm.py::test_e2e_npm[npm_smoketest_lockfile2]
  - tests/integration/test_rpm.py::test_rpm_packages[rpm_unmatched_checksum]
  - tests/integration/test_gomod.py::test_e2e_gomod[gomod_1.18_e2e_test]
  - tests/integration/test_gomod.py::test_gomod_packages[gomod_multiple_modules_missing_checksums]
  - tests/integration/test_npm.py::test_e2e_npm[npm_lockfile3_dev_optional_peer_deps]
  - tests/integration/test_npm.py::test_e2e_npm[npm_lockfile3_multiple_dep_versions]
  - tests/integration/test_rpm.py::test_rpm_packages[rpm_multiple_packages]
  - tests/integration/test_yarn.py::test_e2e_yarn[yarn_e2e_test_multiple_packages]
  - tests/integration/test_gomod.py::test_gomod_packages[gomod_vendor_check_correct_vendor]
  - tests/integration/test_gomod.py::test_gomod_packages[gomod_with_deps]
  - tests/integration/test_rpm.py::test_rpm_packages[rpm_missing_checksums]
  - tests/integration/test_yarn.py::test_yarn_packages[yarn_zero_installs]
  - tests/integration/test_npm.py::test_npm_smoketest[npm_bundled_lockfile3]
  - tests/integration/test_gomod.py::test_gomod_packages[gomod_local_deps]
  - tests/integration/test_yarn.py::test_yarn_packages[yarn_correct_version_installed_by_corepack]
  - tests/integration/test_yarn.py::test_e2e_yarn[yarn_e2e_test]
  - tests/integration/test_rpm.py::test_rpm_packages[rpm_unexpected_size]
  - tests/integration/test_npm.py::test_e2e_npm[npm_multiple_packages_lockfile3]
  - tests/integration/test_gomod.py::test_e2e_gomod[gomod_1.20_e2e_dirty_go.mod]
  - tests/integration/test_npm.py::test_e2e_npm[npm_lockfile3_aliased_deps]
  - tests/integration/test_rpm.py::test_e2e_rpm[rpm_e2e_test]
collected 51 items / 39 deselected / 12 selected    <<< NOTICE THE SAME LINE                                            

tests/integration/test_pip.py
...
```
NB: if you decide to adopt the latter, i.e. not using marks then I think it's worth flipping the boolean logic of `does_not_test_any_affected_pms`.

Reply:
> It looks like some additional moves are required to actually not run them, like adding a skip mark or clearing the list of tests.

You can find all that in my example, you need to modify the items list in-place as that one is passed by reference, that's the reason why you actually ran tests which you marked as deselected. 

>  I like the skip mark approach because it leaves a clear trail in logs -- all skipped tests are skipped with a reason stated right next to test's name, and the reason is "Does not test any modified code"

Well, you can easily tweak that in my example by prepending the same msg string to the "Deselected tests" output et voila achieve the same!

> Currently I am not convinced that traversing a list of ~100 objects is expensive enough to warrant more confusing tox -eintegration output, so I will keep my approach with marks. I am open to changing that though.

Fair enough, I'm not too picky about this and marks can accomplish the same outcome, so let's keep that approach instead of fiddling with test deselection.

Reply:
This use case is exactly what 'skip' marks are for.

Reply:
> This use case is exactly what 'skip' marks are for.

Uhm, I'm not sure, I'd argue that adding the skip marker dynamically during test collection is a clever way to solve a problem rather than a clean and expected solution, but since we're staying within the public API boundaries....Skips are IMO meant to be used directly in the decorator in cases where tests are to be conditionally skipped but you know the condition beforehand (which is not the case here obviously). 
pytest documents the collection process here: https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.hookspec.pytest_collection and they mention deselection, but adding skip marks does look a little less hacky and has a clear output, so why not.

Interestingly enough usage of `pytest.skip` directly is discouraged although I admit I don't understand why it's okay during test collection when you use that particular flag: << This is out of scope though since we moved away from that approach already

https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-skip
> This function should be called only during testing (setup, call or teardown) or during collection by using the allow_module_level flag.